File: R/03_analysis.R
# NBA jersey numbers and player heights over time --- R / tidyverse analysis
#
# Reads the harvested roster table and produces:
#   - output/results_r.csv   season-level metrics (reconciled against the Python run)
#   - output/models_r.csv    fitted model coefficients
#   - figures/*.png          publication figures
#
# The Python script python/03_analysis.py computes the same quantities
# independently; python/04_reconcile.py checks that the two agree. That
# cross-implementation check is the point: it catches cleaning mistakes that a
# single implementation would silently carry through.
#
# Run:  Rscript R/03_analysis.R

suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(tidyr)
  library(stringr)
  library(purrr)
  library(ggplot2)
  library(broom)
})

ROOT <- normalizePath(file.path(dirname(sub("^--file=", "", grep("^--file=", commandArgs(), value = TRUE)[1])), ".."))
if (is.na(ROOT) || !dir.exists(ROOT)) ROOT <- normalizePath(".")

DATA_CSV <- file.path(ROOT, "data", "nba_rosters.csv")
OUT_DIR <- file.path(ROOT, "output")
FIG_DIR <- file.path(ROOT, "figures")
dir.create(OUT_DIR, showWarnings = FALSE, recursive = TRUE)
dir.create(FIG_DIR, showWarnings = FALSE, recursive = TRUE)

# ---- Palette (validated for CVD separation; see README) ---------------------
PAL <- c(blue = "#2a78d6", orange = "#eb6834", aqua = "#1baf7a")
INK <- "#0b0b0b"; INK2 <- "#52514e"; MUTED <- "#898781"
GRID <- "#e1e0d9"; AXIS <- "#c3c2b7"; SURFACE <- "#fcfcfb"

theme_study <- function() {
  theme_minimal(base_size = 11) +
    theme(
      plot.background = element_rect(fill = SURFACE, colour = NA),
      panel.background = element_rect(fill = SURFACE, colour = NA),
      panel.grid.major = element_line(colour = GRID, linewidth = 0.3),
      panel.grid.minor = element_blank(),
      axis.line = element_line(colour = AXIS, linewidth = 0.3),
      axis.text = element_text(colour = MUTED, size = 8.5),
      axis.title = element_text(colour = INK2, size = 9),
      plot.title = element_text(colour = INK, face = "bold", size = 12),
      plot.subtitle = element_text(colour = INK2, size = 9),
      plot.caption = element_text(colour = MUTED, size = 7.5, hjust = 0),
      legend.position = "none"   # series are direct-labelled instead
    )
}

# ---- Load and clean ---------------------------------------------------------
# Documented cleaning decisions:
#   * HEIGHT arrives as "6-9"; parsed to inches. Rows without height are dropped.
#   * NUM is character so "00" stays distinct from "0"; blanks/NA are dropped
#     for number-based metrics but retained for height metrics.
#   * POSITION is "G"/"F"/"C" in early seasons and "G-F" etc. later; the leading
#     letter is used so the grouping is comparable across the whole span.
#   * A player traded mid-season appears on two rosters; the first row per
#     (PLAYER_ID, season) is kept so players are not double counted.

# Pure functions live in functions.R so the test suite (tests/R) can source
# them without running this pipeline.
source(file.path(ROOT, "R", "functions.R"))

raw <- read_csv(DATA_CSV, col_types = cols(.default = col_character()), progress = FALSE)

players <- raw %>%
  mutate(
    season_start = as.integer(season_start),
    height_in = parse_height_in(HEIGHT),
    jersey_raw = str_trim(NUM),
    jersey_num = suppressWarnings(as.numeric(jersey_raw)),
    position_group = str_sub(POSITION, 1, 1),
    weight_lb = suppressWarnings(as.numeric(WEIGHT))
  ) %>%
  filter(!is.na(height_in)) %>%
  distinct(PLAYER_ID, season_start, .keep_all = TRUE)

message(sprintf("Loaded %s player-seasons across %d seasons (%d-%d)",
                format(nrow(players), big.mark = ","),
                dplyr::n_distinct(players$season_start),
                min(players$season_start), max(players$season_start)))

# ---- Season-level metrics ---------------------------------------------------
# Metric definitions:
#   mean_height_in    mean listed height, inches
#   share_above_55    share of numbered players wearing a number above 55.
#                     Numbers above 55 were historically unusual; the league's
#                     referee hand-signal convention is built on digits 0-5.
#   share_zero        share wearing "0" or "00"
#   jersey_height_cor Pearson correlation between jersey number and height
#   num6_count        players wearing 6 (retired league-wide for Bill Russell
#                     ahead of 2022-23, with existing wearers grandfathered)

numbered <- players %>% filter(!is.na(jersey_num))

height_by_season <- players %>%
  group_by(season_start) %>%
  summarise(
    n = n(),
    value = mean(height_in),
    sd = sd(height_in),
    .groups = "drop"
  ) %>%
  mutate(
    se = sd / sqrt(n),
    ci_lo = value - qt(0.975, n - 1) * se,
    ci_hi = value + qt(0.975, n - 1) * se,
    metric = "mean_height_in"
  )

number_metrics <- numbered %>%
  group_by(season_start) %>%
  summarise(
    n = n(),
    share_above_55 = mean(jersey_num > 55),
    share_zero = mean(jersey_raw %in% c("0", "00")),
    num6_count = sum(jersey_num == 6),
    jersey_height_cor = cor(jersey_num, height_in),
    .groups = "drop"
  ) %>%
  pivot_longer(
    c(share_above_55, share_zero, num6_count, jersey_height_cor),
    names_to = "metric", values_to = "value"
  )

results <- bind_rows(
  height_by_season %>% select(metric, season_start, value, n),
  number_metrics %>% select(metric, season_start, value, n)
) %>%
  arrange(metric, season_start)

write_csv(results, file.path(OUT_DIR, "results_r.csv"))

# ---- Models -----------------------------------------------------------------
# 1) Height trend. A single straight line across four decades would be the wrong
#    model: heights rose then flattened. Fit a piecewise-linear trend with a knot
#    at 1990 so the two eras get their own slope, and report both.
h <- height_by_season %>% mutate(post = pmax(season_start - 1990, 0))
m_height <- lm(value ~ season_start + post, data = h, weights = n)

# 1b) Regime-aware model: piecewise trend with knots at 1990 and the 2002 peak,
#     plus a LEVEL SHIFT at the 2019-20 measurement rule change. The shift term
#     turns the eyeballed "-0.61 in step" into a modelled estimate with a CI,
#     and keeps the trend slopes from being contaminated by the break.
#     (Fit comparison on this data: R^2 0.85 vs 0.30 for the single-knot model.)
hr <- height_by_season %>% mutate(
  k1990 = pmax(season_start - 1990, 0),
  k2002 = pmax(season_start - 2002, 0),
  shift2019 = as.numeric(season_start >= 2019)
)
m_regime <- lm(value ~ season_start + k1990 + k2002 + shift2019, data = hr, weights = n)

# 2) Has the jersey-number/height association weakened? Regress the per-season
#    correlation on season.
cors <- number_metrics %>% filter(metric == "jersey_height_cor")
m_cor <- lm(value ~ season_start, data = cors, weights = n)

models <- bind_rows(
  tidy(m_height, conf.int = TRUE) %>% mutate(model = "height_piecewise"),
  tidy(m_regime, conf.int = TRUE) %>% mutate(model = "height_regime"),
  tidy(m_cor, conf.int = TRUE) %>% mutate(model = "jersey_height_cor_trend")
) %>%
  select(model, term, estimate, std.error, statistic, p.value, conf.low, conf.high)

write_csv(models, file.path(OUT_DIR, "models_r.csv"))

# 2b) Within-player year-over-year height change: the composition-free check on
#     the 2019-20 measurement step. The aggregate mean can move because WHO
#     plays changed; a continuing player's listed height only moves when the
#     measurement itself does. Median offseason churn ~2%; the 2019 break 61%.
#     (Implementation in R/functions.R, unit-tested.)
write_csv(within_player_changes(players), file.path(OUT_DIR, "within_player_r.csv"))

# 3) Uncertainty without a distributional assumption: bootstrap the difference in
#    the jersey/height correlation between the first and last five seasons.
set.seed(2026)
era_boot_diff <- function(reps = 2000) {
  seasons <- sort(unique(numbered$season_start))
  # Windows must not overlap or the "change" is partly a season compared to itself.
  if (length(seasons) < 12) {
    message("Fewer than 12 seasons available; skipping era bootstrap.")
    return(NULL)
  }
  first5 <- head(seasons, 5)
  last5 <- tail(seasons, 5)
  a <- numbered %>% filter(season_start %in% first5) %>% select(jersey_num, height_in)
  b <- numbered %>% filter(season_start %in% last5) %>% select(jersey_num, height_in)
  if (nrow(a) < 30 || nrow(b) < 30) return(NULL)
  map_dbl(seq_len(reps), function(i) {
    ia <- sample.int(nrow(a), replace = TRUE)
    ib <- sample.int(nrow(b), replace = TRUE)
    cor(b$jersey_num[ib], b$height_in[ib]) - cor(a$jersey_num[ia], a$height_in[ia])
  })
}
boot <- era_boot_diff()
if (!is.null(boot)) {
  ci <- quantile(boot, c(0.025, 0.975))
  write_csv(
    tibble(quantity = "cor_diff_last5_minus_first5",
           estimate = mean(boot), ci_lo = ci[[1]], ci_hi = ci[[2]], reps = length(boot)),
    file.path(OUT_DIR, "bootstrap_r.csv")
  )
  message(sprintf("Bootstrap correlation change: %.3f [%.3f, %.3f]",
                  mean(boot), ci[[1]], ci[[2]]))
}

# ---- Figures ----------------------------------------------------------------
n_seasons <- dplyr::n_distinct(players$season_start)

# Figure 1: mean height with confidence band. One series, so no legend.
p1 <- ggplot(height_by_season, aes(season_start, value)) +
  geom_ribbon(aes(ymin = ci_lo, ymax = ci_hi), fill = PAL[["blue"]], alpha = 0.18) +
  geom_line(colour = PAL[["blue"]], linewidth = 0.7) +
  # The single largest year-over-year move in the series is a rule change, not a
  # basketball trend. Marking it keeps the chart from telling a false story.
  geom_vline(xintercept = 2019, colour = PAL[["orange"]], linewidth = 0.5, linetype = "dotted") +
  annotate("text", x = 2018.4, y = max(height_by_season$ci_hi), hjust = 1, vjust = 1,
           label = "2019-20: measured heights,\nno shoes, required",
           colour = PAL[["orange"]], size = 2.9) +
  labs(
    title = "Most of the height “decline” is a 2019 measurement change",
    subtitle = "Mean listed height per season, with 95% confidence band",
    x = NULL, y = "Height (inches)",
    caption = "Source: stats.nba.com CommonTeamRoster. One row per player-season; mid-season trades de-duplicated."
  ) +
  theme_study()
ggsave(file.path(FIG_DIR, "fig1_height_trend_r.png"), p1, width = 7.5, height = 4.2, dpi = 200)

# Figure 2: height by position group, direct-labelled (no legend box).
pos <- players %>%
  filter(position_group %in% c("G", "F", "C")) %>%
  group_by(season_start, position_group) %>%
  summarise(value = mean(height_in), n = n(), .groups = "drop")

pos_cols <- c(G = PAL[["blue"]], F = PAL[["orange"]], C = PAL[["aqua"]])
pos_labels <- pos %>% group_by(position_group) %>% filter(season_start == max(season_start)) %>% ungroup()

p2 <- ggplot(pos, aes(season_start, value, colour = position_group)) +
  geom_line(linewidth = 0.7) +
  geom_text(
    data = pos_labels,
    aes(label = c(G = "Guards", F = "Forwards", C = "Centers")[position_group]),
    hjust = -0.1, size = 3, fontface = "bold", show.legend = FALSE
  ) +
  scale_colour_manual(values = pos_cols) +
  scale_x_continuous(expand = expansion(mult = c(0.02, 0.14))) +
  labs(
    title = "Guards grew taller as the center–guard gap narrowed",
    subtitle = "Mean listed height per season by position group",
    x = NULL, y = "Height (inches)",
    caption = "Position taken from the leading letter so early 'G'/'F'/'C' and later 'G-F' codes are comparable."
  ) +
  theme_study()
ggsave(file.path(FIG_DIR, "fig2_height_by_position_r.png"), p2, width = 7.5, height = 4.2, dpi = 200)

# Figure 3: jersey number conventions over time.
num_share <- number_metrics %>%
  filter(metric %in% c("share_above_55", "share_zero")) %>%
  mutate(label = if_else(metric == "share_above_55", "Number above 55", "Wearing 0 or 00"))

lab3 <- num_share %>% group_by(metric) %>% filter(season_start == max(season_start)) %>% ungroup()

p3 <- ggplot(num_share, aes(season_start, value, colour = metric)) +
  geom_line(linewidth = 0.7) +
  geom_text(data = lab3, aes(label = label), hjust = -0.05, size = 3, fontface = "bold") +
  scale_colour_manual(values = c(share_above_55 = PAL[["blue"]], share_zero = PAL[["orange"]])) +
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  scale_x_continuous(expand = expansion(mult = c(0.02, 0.20))) +
  labs(
    title = "Jersey number conventions have loosened",
    subtitle = "Share of numbered players per season",
    x = NULL, y = NULL,
    caption = "Numbers above 55 sit outside the digit range the NBA's referee hand-signal convention is built on."
  ) +
  theme_study()
ggsave(file.path(FIG_DIR, "fig3_number_conventions_r.png"), p3, width = 7.5, height = 4.2, dpi = 200)

# Figure 4: jersey/height association per season.
p4 <- ggplot(cors, aes(season_start, value)) +
  geom_hline(yintercept = 0, colour = AXIS, linewidth = 0.3) +
  geom_line(colour = PAL[["blue"]], linewidth = 0.7) +
  labs(
    title = "The link between jersey number and height has faded",
    subtitle = "Per-season correlation between jersey number and listed height",
    x = NULL, y = "Pearson correlation",
    caption = "Positive values mean taller players tend to wear higher numbers."
  ) +
  theme_study()
ggsave(file.path(FIG_DIR, "fig4_jersey_height_cor_r.png"), p4, width = 7.5, height = 4.2, dpi = 200)

message("R analysis complete: ", n_seasons, " seasons; wrote results, models, and 4 figures.")

File: python/03_analysis.py
"""NBA jersey numbers and player heights over time --- Python analysis (polars + plotly).

Deliberately an independent implementation of R/03_analysis.R rather than a port:
same metric definitions, written from the definitions rather than from the R code,
so that 04_reconcile.py is a real check instead of a formality.

Outputs:
    output/results_python.csv     season-level metrics
    output/models_python.csv      fitted model coefficients
    output/bootstrap_python.csv   bootstrap CI for the era change in correlation
    figures/*_py.png              static figures
    figures/*_py.html             interactive versions (hover tooltips)

statsmodels is used for the weighted regressions; there is no polars-native
equivalent, and it is what gives standard errors and confidence intervals
comparable to R's lm().

Run:  python python/03_analysis.py
"""
from __future__ import annotations

from pathlib import Path

import numpy as np
import plotly.graph_objects as go
import polars as pl
import statsmodels.api as sm
from scipy import stats

ROOT = Path(__file__).resolve().parents[1]
DATA_CSV = ROOT / "data" / "nba_rosters.csv"
OUT_DIR = ROOT / "output"
FIG_DIR = ROOT / "figures"
OUT_DIR.mkdir(exist_ok=True)
FIG_DIR.mkdir(exist_ok=True)

# Palette validated for colour-vision separation (see README).
BLUE, ORANGE, AQUA = "#2a78d6", "#eb6834", "#1baf7a"
INK, INK2, MUTED = "#0b0b0b", "#52514e", "#898781"
GRID, AXIS, SURFACE = "#e1e0d9", "#c3c2b7", "#fcfcfb"


def base_layout(title: str, subtitle: str, y_title: str | None) -> go.Layout:
    """Recessive chrome: hairline grid, no top/right spines, muted ticks."""
    return go.Layout(
        title=dict(text=f"<b>{title}</b><br><span style='font-size:12px;color:{INK2}'>"
                        f"{subtitle}</span>",
                   font=dict(size=17, color=INK), x=0, xanchor="left"),
        paper_bgcolor=SURFACE,
        plot_bgcolor=SURFACE,
        font=dict(family="system-ui, -apple-system, Segoe UI, sans-serif",
                  size=12, color=INK2),
        xaxis=dict(showgrid=True, gridcolor=GRID, gridwidth=0.5, zeroline=False,
                   linecolor=AXIS, linewidth=1, tickfont=dict(color=MUTED, size=11)),
        yaxis=dict(title=dict(text=y_title, font=dict(color=INK2, size=12)) if y_title else None,
                   showgrid=True, gridcolor=GRID, gridwidth=0.5, zeroline=False,
                   linecolor=AXIS, linewidth=1, tickfont=dict(color=MUTED, size=11)),
        showlegend=False,          # series are direct-labelled instead
        hovermode="x unified",
        margin=dict(l=70, r=130, t=80, b=50),
        width=900, height=500,
    )


def save(fig: go.Figure, stem: str) -> None:
    fig.write_image(FIG_DIR / f"{stem}.png", scale=2)
    fig.write_html(FIG_DIR / f"{stem}.html", include_plotlyjs="cdn")


def parse_height_expr(col: str) -> pl.Expr:
    """'6-9' -> 81.0. Anything unparseable becomes null."""
    feet = pl.col(col).str.extract(r"^(\d+)-", 1).cast(pl.Float64, strict=False)
    inches = pl.col(col).str.extract(r"-(\d+)$", 1).cast(pl.Float64, strict=False)
    return (feet * 12 + inches).alias("height_in")


def load_players() -> pl.DataFrame:
    """Load and clean. Cleaning decisions are documented in the README."""
    raw = pl.read_csv(DATA_CSV, infer_schema_length=0)
    df = (
        raw.with_columns(
            pl.col("season_start").cast(pl.Int32),
            parse_height_expr("HEIGHT"),
            pl.col("NUM").fill_null("").str.strip_chars().alias("jersey_raw"),
            pl.col("POSITION").fill_null("").str.slice(0, 1).alias("position_group"),
        )
        .with_columns(
            pl.col("jersey_raw").cast(pl.Float64, strict=False).alias("jersey_num")
        )
        .filter(pl.col("height_in").is_not_null())
        # A mid-season trade puts one player on two rosters; keep the first row.
        .unique(subset=["PLAYER_ID", "season_start"], keep="first", maintain_order=True)
    )
    return df


def season_metrics(players: pl.DataFrame) -> pl.DataFrame:
    numbered = players.filter(pl.col("jersey_num").is_not_null())

    height = (
        players.group_by("season_start")
        .agg(n=pl.len(), value=pl.col("height_in").mean())
        .with_columns(metric=pl.lit("mean_height_in"))
        .select("metric", "season_start", "value", "n")
    )

    nums = (
        numbered.group_by("season_start")
        .agg(
            n=pl.len(),
            share_above_55=(pl.col("jersey_num") > 55).mean(),
            share_zero=pl.col("jersey_raw").is_in(["0", "00"]).mean(),
            num6_count=(pl.col("jersey_num") == 6).sum().cast(pl.Float64),
            jersey_height_cor=pl.corr("jersey_num", "height_in"),
        )
        .unpivot(
            index=["season_start", "n"],
            on=["share_above_55", "share_zero", "num6_count", "jersey_height_cor"],
            variable_name="metric", value_name="value",
        )
        .select("metric", "season_start", "value", "n")
    )

    return pl.concat([height, nums]).sort(["metric", "season_start"])


def fit_models(results: pl.DataFrame) -> pl.DataFrame:
    """Piecewise height trend (knot at 1990), a regime-aware height model, and
    the correlation trend."""
    rows: list[dict] = []

    h = (results.filter(pl.col("metric") == "mean_height_in")
         .with_columns(post=(pl.col("season_start") - 1990).clip(lower_bound=0)))
    X1 = sm.add_constant(h.select("season_start", "post").to_numpy())
    m1 = sm.WLS(h["value"].to_numpy(), X1, weights=h["n"].to_numpy()).fit()

    # Regime-aware model: piecewise trend with knots at 1990 and the 2002 peak,
    # plus a LEVEL SHIFT at the 2019-20 measurement rule change. The shift term
    # turns the eyeballed "-0.61 in step" into a modelled estimate with a CI,
    # and keeps the trend slopes from being contaminated by the break.
    # (Fit comparison on this data: R^2 0.85 vs 0.30 for the single-knot model.)
    hr = h.with_columns(
        k1990=(pl.col("season_start") - 1990).clip(lower_bound=0),
        k2002=(pl.col("season_start") - 2002).clip(lower_bound=0),
        shift2019=(pl.col("season_start") >= 2019).cast(pl.Float64),
    )
    X1b = sm.add_constant(hr.select("season_start", "k1990", "k2002", "shift2019").to_numpy())
    m1b = sm.WLS(hr["value"].to_numpy(), X1b, weights=hr["n"].to_numpy()).fit()

    c = (results.filter(pl.col("metric") == "jersey_height_cor")
         .drop_nulls("value"))
    X2 = sm.add_constant(c.select("season_start").to_numpy())
    m2 = sm.WLS(c["value"].to_numpy(), X2, weights=c["n"].to_numpy()).fit()

    naming = {
        "height_piecewise": (m1, ["(Intercept)", "season_start", "post"]),
        "height_regime": (m1b, ["(Intercept)", "season_start", "k1990", "k2002", "shift2019"]),
        "jersey_height_cor_trend": (m2, ["(Intercept)", "season_start"]),
    }
    for model_name, (model, terms) in naming.items():
        ci = model.conf_int()
        for i, term in enumerate(terms):
            rows.append({
                "model": model_name,
                "term": term,
                "estimate": float(model.params[i]),
                "std.error": float(model.bse[i]),
                "statistic": float(model.tvalues[i]),
                "p.value": float(model.pvalues[i]),
                "conf.low": float(ci[i][0]),
                "conf.high": float(ci[i][1]),
            })
    return pl.DataFrame(rows)


def within_player_changes(players: pl.DataFrame) -> pl.DataFrame:
    """Year-over-year listed-height change for players present in both seasons.

    This is the composition-free check on the 2019-20 measurement step: the
    aggregate mean can move because WHO plays changed, but a continuing player's
    listed height only moves when the measurement itself does. In the median
    offseason ~2% of continuing players' listed heights change (no other
    offseason since 1980 exceeds 27%, and most sit under 5%); if the rule
    change is real it should show up as a majority of continuing players
    'shrinking' in a single offseason.
    """
    seasons = players["season_start"].unique().sort().to_list()
    rows = []
    for s0, s1 in zip(seasons, seasons[1:]):
        a = players.filter(pl.col("season_start") == s0).select("PLAYER_ID", h0="height_in")
        b = players.filter(pl.col("season_start") == s1).select("PLAYER_ID", h1="height_in")
        j = a.join(b, on="PLAYER_ID", how="inner").with_columns(
            delta=pl.col("h1") - pl.col("h0"))
        if j.height == 0:
            continue
        rows.append({
            "pair_start": s0,
            "n_matched": j.height,
            "mean_delta": float(j["delta"].mean()),
            "share_shrunk": float((j["delta"] < 0).mean()),
            "share_same": float((j["delta"] == 0).mean()),
            "share_grew": float((j["delta"] > 0).mean()),
        })
    return pl.DataFrame(rows)


def era_bootstrap(players: pl.DataFrame, reps: int = 2000, seed: int = 2026) -> pl.DataFrame | None:
    """Bootstrap the change in jersey/height correlation, last 5 vs first 5 seasons."""
    numbered = players.filter(pl.col("jersey_num").is_not_null())
    seasons = numbered["season_start"].unique().sort().to_list()
    if len(seasons) < 12:
        print("Fewer than 12 seasons available; skipping era bootstrap.")
        return None

    a = numbered.filter(pl.col("season_start").is_in(seasons[:5])).select(
        "jersey_num", "height_in").to_numpy()
    b = numbered.filter(pl.col("season_start").is_in(seasons[-5:])).select(
        "jersey_num", "height_in").to_numpy()
    if len(a) < 30 or len(b) < 30:
        return None

    rng = np.random.default_rng(seed)
    diffs = np.empty(reps)
    for i in range(reps):
        ia = rng.integers(0, len(a), len(a))
        ib = rng.integers(0, len(b), len(b))
        diffs[i] = (np.corrcoef(b[ib, 0], b[ib, 1])[0, 1]
                    - np.corrcoef(a[ia, 0], a[ia, 1])[0, 1])

    lo, hi = np.quantile(diffs, [0.025, 0.975])
    print(f"Bootstrap correlation change: {diffs.mean():.3f} [{lo:.3f}, {hi:.3f}]")
    return pl.DataFrame([{
        "quantity": "cor_diff_last5_minus_first5",
        "estimate": float(diffs.mean()), "ci_lo": float(lo), "ci_hi": float(hi),
        "reps": reps,
    }])


def make_figures(players: pl.DataFrame, results: pl.DataFrame) -> None:
    # --- Figure 1: mean height with a t-based 95% band. One series, no legend.
    h = (players.group_by("season_start")
         .agg(n=pl.len(), mean=pl.col("height_in").mean(), sd=pl.col("height_in").std())
         .sort("season_start"))
    n = h["n"].to_numpy()
    se = h["sd"].to_numpy() / np.sqrt(n)
    crit = stats.t.ppf(0.975, n - 1)
    x = h["season_start"].to_list()
    mean, lo, hi = h["mean"].to_numpy(), h["mean"].to_numpy() - crit * se, h["mean"].to_numpy() + crit * se

    fig = go.Figure(layout=base_layout(
        "Most of the height &#8220;decline&#8221; is a 2019 measurement change",
        "Mean listed height per season, with 95% confidence band", "Height (inches)"))
    fig.add_trace(go.Scatter(x=x + x[::-1], y=list(hi) + list(lo)[::-1], fill="toself",
                             fillcolor="rgba(42,120,214,0.18)", line=dict(width=0),
                             hoverinfo="skip", showlegend=False))
    fig.add_trace(go.Scatter(x=x, y=mean, mode="lines", line=dict(color=BLUE, width=2),
                             name="Mean height",
                             hovertemplate="%{x}: %{y:.2f} in<extra></extra>"))
    # The single largest year-over-year move in the series is a rule change, not a
    # basketball trend. Marking it keeps the chart from telling a false story.
    if 2019 in x:
        fig.add_vline(x=2019, line=dict(color=ORANGE, width=1.5, dash="dot"))
        fig.add_annotation(x=2019, y=max(hi), yanchor="bottom", xanchor="right", xshift=-6,
                           text="2019-20: measured heights,<br>no shoes, required",
                           showarrow=False, align="right",
                           font=dict(color=ORANGE, size=11))
    save(fig, "fig1_height_trend_py")

    # --- Figure 2: by position, direct-labelled at the right edge.
    pos = (players.filter(pl.col("position_group").is_in(["G", "F", "C"]))
           .group_by("season_start", "position_group")
           .agg(value=pl.col("height_in").mean())
           .sort("season_start"))
    colours = {"G": BLUE, "F": ORANGE, "C": AQUA}
    names = {"G": "Guards", "F": "Forwards", "C": "Centers"}

    fig = go.Figure(layout=base_layout(
        "Guards grew taller as the center–guard gap narrowed",
        "Mean listed height per season by position group", "Height (inches)"))
    for grp in ["G", "F", "C"]:
        sub = pos.filter(pl.col("position_group") == grp).sort("season_start")
        if sub.height == 0:
            continue
        fig.add_trace(go.Scatter(
            x=sub["season_start"].to_list(), y=sub["value"].to_list(), mode="lines",
            line=dict(color=colours[grp], width=2), name=names[grp],
            hovertemplate=f"{names[grp]} %{{x}}: %{{y:.2f}} in<extra></extra>"))
        fig.add_annotation(x=sub["season_start"][-1], y=sub["value"][-1], text=names[grp],
                           showarrow=False, xanchor="left", xshift=8,
                           font=dict(color=colours[grp], size=13))
    save(fig, "fig2_height_by_position_py")

    # --- Figure 3: number conventions.
    fig = go.Figure(layout=base_layout(
        "Jersey number conventions have loosened",
        "Share of numbered players per season", "Share of numbered players"))
    for metric, colour, label in (("share_above_55", BLUE, "Number above 55"),
                                  ("share_zero", ORANGE, "Wearing 0 or 00")):
        sub = results.filter(pl.col("metric") == metric).sort("season_start")
        if sub.height == 0:
            continue
        fig.add_trace(go.Scatter(
            x=sub["season_start"].to_list(), y=sub["value"].to_list(), mode="lines",
            line=dict(color=colour, width=2), name=label,
            hovertemplate=f"{label} %{{x}}: %{{y:.1%}}<extra></extra>"))
        fig.add_annotation(x=sub["season_start"][-1], y=sub["value"][-1], text=label,
                           showarrow=False, xanchor="left", xshift=8,
                           font=dict(color=colour, size=13))
    fig.update_yaxes(tickformat=".0%")
    save(fig, "fig3_number_conventions_py")

    # --- Figure 4: jersey/height correlation.
    cors = results.filter(pl.col("metric") == "jersey_height_cor").sort("season_start")
    if cors.height:
        fig = go.Figure(layout=base_layout(
            "The link between jersey number and height has faded",
            "Per-season correlation between jersey number and listed height",
            "Pearson correlation"))
        fig.add_hline(y=0, line=dict(color=AXIS, width=1))
        fig.add_trace(go.Scatter(
            x=cors["season_start"].to_list(), y=cors["value"].to_list(), mode="lines",
            line=dict(color=BLUE, width=2), name="Correlation",
            hovertemplate="%{x}: r = %{y:.3f}<extra></extra>"))
        save(fig, "fig4_jersey_height_cor_py")


def main() -> int:
    players = load_players()
    seasons = players["season_start"]
    print(f"Loaded {players.height:,} player-seasons across {seasons.n_unique()} seasons "
          f"({seasons.min()}-{seasons.max()})")

    results = season_metrics(players)
    results.write_csv(OUT_DIR / "results_python.csv")

    fit_models(results).write_csv(OUT_DIR / "models_python.csv")

    within_player_changes(players).write_csv(OUT_DIR / "within_player_python.csv")

    boot = era_bootstrap(players)
    if boot is not None:
        boot.write_csv(OUT_DIR / "bootstrap_python.csv")

    make_figures(players, results)
    print(f"Python analysis complete: {seasons.n_unique()} seasons; "
          f"wrote results, models, and 4 figures (PNG + interactive HTML).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
